cartography/intel/rapid7/util.py

```diff
@@ -0,0 +1,303 @@
+    return instance_id
+
+
+def extract_rapid7_history_lastseen(history):
+    """
+    Extract from rapid7 history, last seen/scan time
+    """
+    lastseen = history[0]["date"]
+    for entry in history:
+        if lastseen < entry["date"]:
+            lastseen = entry["date"]
+    return lastseen
+
+
+def extract_rapid7_history_firstseen(history):
+    """
+    Extract from rapid7 history, first seen/scan time
+    """
+    firstseen = history[0]["date"]
+    for entry in history:
+        if firstseen > entry["date"]:
+            firstseen = entry["date"]
+    return firstseen
+
+
+# pylint: disable=too-many-arguments,too-many-locals
+def Rapid7Hosts(
+    authorization: tuple[str, str, str, bool],
+    limit: int = 10000,
+    page: int = 0,
+    sort: str = "",
+    nexpose_timeout: int = 5,
+) -> array:
+    """
+    Get Nexpose inventory full data to array
+    https://help.rapid7.com/insightvm/en-us/api/index.html#operation/getAssets
+
+    To filter, use https://help.rapid7.com/insightvm/en-us/api/index.html#operation/findAssets
+
+    Warning! if very large assets count (>10k, likely depending on assets details), it
+    will have impact on memory needed for processing and could trigger linux oom-killer.
+    Possible option: have the server generate a report and just retrieve it.
+    https://help.rapid7.com/insightvm/en-us/api/index.html#operation/downloadReport
+    """
+    (
+        nexpose_user,
+        nexpose_password,
+        nexpose_server_url,
+        nexpose_verify_cert,
+    ) = authorization
+    url = nexpose_server_url + "/api/3/assets"
+    headers = {"Content-type": "application/json", "Accept": "application/json"}
+
+    logger.info("nexpose_verify_cert: %s", nexpose_verify_cert)
+    if nexpose_verify_cert is not True:
+        nexpose_verify_cert = False
+        logger.warning("Requested to not verify certificate (%s)", nexpose_verify_cert)
+
+    count = total_resources = 0
+    size_interval = 500
+    result_array = []
+    while count < limit:
+        logger.info(
+            "count %s, page %s, total_resources %s",
+            count,
+            page,
+            total_resources,
+        )
+        params = {"page": page, "size": size_interval, "sort": sort}
+        # logger.debug("GetAssets params %s", params)
+        try:
+            resp = requests.get(
+                url,
+                headers=headers,
+                auth=(nexpose_user, nexpose_password),
+                verify=nexpose_verify_cert,
+                params=params,
+                timeout=nexpose_timeout,
+            )
+            resp.raise_for_status()
+
+            data = resp.json()
+
+            total_resources = data["page"]["totalResources"]
+            if total_resources < limit:
+                limit = total_resources
+
+        except requests.HTTPError as exception:
+            logger.warning("GetAssets0: %s", resp.content)
+            logger.exception("GetAssets0: %s", exception)
```

Comment:
Do you mean to have both of these log messages?

Reply:
rapid7 API content is usually huge. that's one reason that I split it from exception message itself.

Reply:
input?

Reply:
What kinds of situations would cause this HTTPError? Are you only meaning to cover regular network errors? If so, please remove this. But if you must log an exception, try to come up with a better message, rather than logging the entire data object.

Reply:
removed resp.content